lib/DxilContainer/DxilContainerAssembler.cpp

```diff
@@ -1373,6 +1388,78 @@ class DxilRDATWriter : public DxilPartWriter {
     }
   }
 
+  void UpdatePayloadInfo(const DxilModule &DM) {
+      Module* module = DM.GetModule();
+      assert(module);
+      DataLayout DL(module);
+
+      const DxilTypeSystem& DTS = DM.GetTypeSystem();
+      int TypeID = 0;
+      int FieldID = 0;
+      for (const auto& entry : DTS.GetPayloadAnnotationMap()) {
+        const StructType* type = entry.first;
+        const auto& annotation = entry.second;
+
+        RuntimeDataPayloadTypeInfo info; 
+        info.ID = TypeID;
+        info.Name = m_pStringBufferPart->Insert(type->getName());
+        m_pPayloadTypeTable->Insert(info);
+
+        for (unsigned i = 0; i < annotation->GetNumFields(); ++i) {
+            Type* fieldType = type->getElementType(i);
+            uint32_t size = 0;
+            bool fieldIsStructType = false;
+
+            if (fieldType->isAggregateType()) {
+              if (fieldType->isStructTy()) {
+                  size = DL.getTypeAllocSize(fieldType);
+                  fieldIsStructType = true;
+              } else {
+                ArrayType *arrayType = cast<ArrayType>(fieldType);
+                fieldType = arrayType->getElementType();
+                size = arrayType->getNumElements();
+                if (fieldType->isStructTy()) {
+                    size *= DL.getTypeAllocSize(fieldType);
+                    fieldIsStructType = true;
+                }
+              }
+            } else if (fieldType->isVectorTy()) {
+              VectorType *vectorType = cast<VectorType>(fieldType);
+              fieldType = vectorType->getElementType();
+              size = vectorType->getNumElements();
+            }
+            
+            RuntimeDataPayloadFieldInfo fieldInfo; 
+
+            const auto& fieldAnnotation = annotation->GetFieldAnnotation(i);
+
+            fieldInfo.ID = FieldID; 
+            fieldInfo.StructTypeName = 0;
+            if (!fieldIsStructType)
+              fieldInfo.Type = static_cast<uint32_t>(
+                  fieldAnnotation.GetCompType().GetKind());
+            else {
+              fieldInfo.Type =
+                  0xffffffffu; // indicates a struct type in the payload
+              fieldInfo.StructTypeName =
+                  m_pStringBufferPart->Insert(cast<StructType>(fieldType)->getName());
```

Comment:
What use is this?  What happens with nested payload structures?  How does this encode fields inside that structure?  What is the PayloadAccessQualifier value for a nested payload struct field?  Is someone expected to look up the struct by name and see that it's a payload structure and that the actual qualifiers are there?  It would be better if we had a nested struct ID for struct type, then Size would always be ElementCount (array count with nested struct).  What about matrix, which is still a StructType, but should be encoded as component type and count?